nova_vm/src/ecmascript/builtins/indexed_collections/typed_array_objects/abstract_operations.rs

```diff
@@ -0,0 +1,686 @@
+    // c. Set obj.[[ArrayLength]] to 0.
+
+    if let Some(length) = length {
+        // 8. Else,
+        // a. Perform ? AllocateTypedArrayBuffer(obj, length).
+        allocate_typed_array_buffer::<T>(agent, obj, length)?;
+    }
+
+    // 9. Return obj.
+    Ok(obj)
+}
+
+/// ### [23.2.5.1.2 InitializeTypedArrayFromTypedArray ( O, srcArray )](https://tc39.es/ecma262/#sec-initializetypedarrayfromtypedarray)
+///
+/// The abstract operation InitializeTypedArrayFromTypedArray takes arguments O
+/// (a TypedArray) and srcArray (a TypedArray) and returns either a normal
+/// completion containing unused or a throw completion.
+pub(crate) fn initialize_typed_array_from_typed_array<T: Viewable, U: Viewable>(
+    agent: &mut Agent,
+    mut gc: GcScope<'_, '_>,
+    o: TypedArray,
+    src_array: TypedArray,
+) -> JsResult<()> {
+    let src_heap_data = &agent[src_array];
+
+    // 1. Let srcData be srcArray.[[ViewedArrayBuffer]].
+    let src_data = src_heap_data.viewed_array_buffer;
+
+    // 2. Let elementType be TypedArrayElementType(O).
+    // 3. Let elementSize be TypedArrayElementSize(O).
+    let element_size = size_of::<T>();
+
+    // 4. Let srcType be TypedArrayElementType(srcArray).
+    // 5. Let srcElementSize be TypedArrayElementSize(srcArray).
+    let src_element_size = size_of::<U>();
+
+    // 6. Let srcByteOffset be srcArray.[[ByteOffset]].
+    let src_byte_offset = src_array.byte_offset(agent);
+
+    // 7. Let srcRecord be MakeTypedArrayWithBufferWitnessRecord(srcArray, seq-cst).
+    let src_record =
+        make_typed_array_with_buffer_witness_record(agent, src_array, Ordering::SeqCst);
+
+    // 8. If IsTypedArrayOutOfBounds(srcRecord) is true, throw a TypeError exception.
+    if is_typed_array_out_of_bounds::<U>(agent, &src_record) {
+        return Err(agent.throw_exception_with_static_message(
+            ExceptionType::TypeError,
+            "TypedArray out of bounds",
+        ));
+    }
+
+    // 9. Let elementLength be TypedArrayLength(srcRecord).
+    let element_length = typed_array_length::<T>(agent, &src_record);
+
+    // 10. Let byteLength be elementSize × elementLength.
+    let byte_length = element_size * element_length;
+
+    // 11. If elementType is srcType, then
+    let data = if T::PROTO == U::PROTO {
+        // a. Let data be ? CloneArrayBuffer(srcData, srcByteOffset, byteLength).
+        clone_array_buffer(agent, src_data, src_byte_offset, byte_length)?
+    } else {
+        // 12. Else,
+        // a. Let data be ? AllocateArrayBuffer(%ArrayBuffer%, byteLength).
+        let array_buffer_constructor = agent.current_realm().intrinsics().array_buffer();
+        let data = allocate_array_buffer(
+            agent,
+            array_buffer_constructor.into_function(),
+            byte_length as u64,
+            None,
+        )?;
+
+        // b. If srcArray.[[ContentType]] is not O.[[ContentType]], throw a TypeError exception.
+        if T::IS_BIGINT != U::IS_BIGINT {
```

Comment:
**nitpick:** We don't need to allocate the array buffer to perform this check.

Reply:
Couldn't errors technically be thrown in the wrong order because `allocate_array_buffer` fails before we come to this check?

Reply:
Oh, you're absolutely right. Nevermind :)